File: src/run_morpho_cli.cpp
// This file is part of MorphoDiTa <http://github.com/ufal/morphodita/>.
//
// Copyright 2015 Institute of Formal and Applied Linguistics, Faculty of
// Mathematics and Physics, Charles University in Prague, Czech Republic.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

#include "morpho/morpho.h"
#include "tagger/tagger.h"
#include "utils/iostreams.h"
#include "utils/options.h"
#include "utils/split.h"
#include "version/version.h"

using namespace ufal::morphodita;

int main(int argc, char* argv[]) {
  iostreams_init();

  options::map options;
  if (!options::parse({{"from_tagger",options::value::none},
                       {"version", options::value::none},
                       {"help", options::value::none}}, argc, argv, options) ||
      options.count("help") ||
      (argc < 2 && !options.count("version")))
    runtime_failure("Usage: " << argv[0] << " [options] dict_file\n"
                    "Options: --from_tagger\n"
                    "         --version\n"
                    "         --help");
  if (options.count("version"))
    return cout << version::version_and_copyright() << endl, 0;

  unique_ptr<morpho> morpho;
  unique_ptr<tagger> tagger;
  if (options.count("from_tagger")) {
    cerr << "Loading dictionary from tagger: ";
    tagger.reset(tagger::load(argv[1]));
    if (!tagger) runtime_failure("Cannot load tagger from file '" << argv[1] << "'!");
  } else {
    cerr << "Loading dictionary: ";
    morpho.reset(morpho::load(argv[1]));
    if (!morpho) runtime_failure("Cannot load dictionary from file '" << argv[1] << "'!");
  }
  cerr << "done" << endl;
  auto& dictionary = options.count("from_tagger") ? *tagger->get_morpho() : *morpho;

  string line;
  vector<string> tokens;
  while (getline(cin, line)) {
    split(line, '\t', tokens);
    if /* analyze */ (tokens.size() == 1) {
      vector<tagged_lemma> lemmas;
      auto result = dictionary.analyze(tokens[0], morpho::GUESSER, lemmas);

      string guesser_name = result == morpho::GUESSER ? "Guesser " : "";
      for (auto&& lemma : lemmas)
        cout << guesser_name << "Lemma: " << lemma.lemma << ' ' << lemma.tag << '\n';
      cout << flush;
    } else if /* generate */ (tokens.size() == 2) {
      vector<tagged_lemma_forms> forms;
      auto result = dictionary.generate(tokens[0], tokens[1].c_str(), morpho::GUESSER, forms);

      string guesser_name = result == morpho::GUESSER ? "Guesser " : "";
      for (auto&& lemma : forms) {
        cout << guesser_name << "Lemma: " << lemma.lemma << '\n';
        for (auto&& form : lemma.forms)
          cout << "  " << form.form << ' ' << form.tag << '\n';
      }
      cout << flush;
    }
  }

  return 0;
}

File: bindings/java/examples/RunMorphoCli.java
// This file is part of MorphoDiTa <http://github.com/ufal/morphodita/>.
//
// Copyright 2015 Institute of Formal and Applied Linguistics, Faculty of
// Mathematics and Physics, Charles University in Prague, Czech Republic.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import cz.cuni.mff.ufal.morphodita.*;
import java.util.Scanner;

class RunMorphoCli {
  public static void main(String[] args) {
    if (args.length < 1) {
      System.err.println("Usage: RunMorphoCli dict_file");
      System.exit(1);
    }

    System.err.print("Loading dictionary: ");
    Morpho morpho = Morpho.load(args[0]);
    if (morpho == null) {
      System.err.println("Cannot load dictionary from file '" + args[0] + "'");
      System.exit(1);
    }
    System.err.println("done");

    TaggedLemmas lemmas = new TaggedLemmas();
    TaggedLemmasForms lemmas_forms = new TaggedLemmasForms();
    Scanner reader = new Scanner(System.in);
    while (reader.hasNextLine()) {
      String[] tokens = reader.nextLine().split("\t", -1);

      if (tokens.length == 1) /* analyze */ {
        int result = morpho.analyze(tokens[0], morpho.GUESSER, lemmas);

        String guesser = result == morpho.GUESSER ? "Guesser " : "";
        for (int i = 0; i < lemmas.size(); i++) {
          TaggedLemma lemma = lemmas.get(i);
          System.out.printf("%sLemma: %s %s\n", guesser, lemma.getLemma(), lemma.getTag());
        }
      } else if (tokens.length == 2) /* generate */ {
        int result = morpho.generate(tokens[0], tokens[1], morpho.GUESSER, lemmas_forms);
        String guesser = result == morpho.GUESSER ? "Guesser " : "";
        for (int i = 0; i < lemmas_forms.size(); i++) {
          TaggedLemmaForms lemma_forms = lemmas_forms.get(i);
          System.out.printf("%sLemma: %s\n", guesser, lemma_forms.getLemma());

          TaggedForms forms = lemma_forms.getForms();
          for (int j = 0; j < forms.size(); j++) {
            TaggedForm form = forms.get(j);
            System.out.printf("  %s %s\n", form.getForm(), form.getTag());
          }
        }
      }
    }
  }
}
